DataStructures/BinarySearchTree/BinarySearchTreeNode.cs

```diff
@@ -0,0 +1,38 @@
+﻿using System;
+
+namespace DataStructures.BinarySearchTree
+{
+    /// <summary>
+    /// Generic node class for BinarySearchTree.
+    /// Keys for each node are immutable.
+    /// </summary>
+    /// <typeparam name="TKey">Type of key for the node. Keys must implement IComparable.</typeparam>
+    public class BinarySearchTreeNode<TKey> where TKey : IComparable<TKey>
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BinarySearchTreeNode{TKey}"/> class.
+        /// </summary>
+        /// <param name="key">The key of this node.</param>
+        public BinarySearchTreeNode(TKey key) => Key = key;
+
+        /// <summary>
+        /// Gets the key for this node.
+        /// </summary>
+        public TKey Key { get; }
+
+        /// <summary>
+        /// Gets or sets the reference to the parent node for this node.
+        /// </summary>
+        public BinarySearchTreeNode<TKey>? Parent { get; set; }
```

Comment:
Can we get away without this property? It seems unnecessary because it duplicates already present information (nodes are recorded as left or right children and this implies their parent). Just my thoughts.